File: Stack/dynamic/ts/Stack.ts
import { Node } from "./Node";

export class Stack {
    head: Node | undefined;

    constructor() {
        this.head = undefined;
    }

    is_empty() {
        return !this.head
    }

    push(node: Node) {
        if (this.is_empty() || !this.head) return;
        node.next = this.head;
        this.head = node;
    }

    pop(): Node | undefined {
        const node = this.head;
        if(!node) return undefined;
        this.head = node.next;
        return node;
    }

    peek(): Node | undefined {
        return this.head;
    }
}

File: src-py/Stack/dynamic/Stack.py
import sys
sys.path.append("../../utils")

from Node import Node

class Stack:
    def __init__(self) -> None:
        self.head = None
        
    def __del__(self) -> None:
        self.head = None
        
    def __str__(self) -> str:
        str_stack = "\tHEAD -> "
        aux_node = self.head
        while (aux_node != None):
            str_stack += f"{aux_node} -> "
            aux_node = aux_node.next
        str_stack += "NULL\n"
        return str_stack
        
    def is_empty(self) -> bool:
        return self.head == None
    
    def push(self, value: Node) -> None:
        value.next = self.head
        self.head = value
        
    def pop(self) -> Node:
        try:
            if self.is_empty():
                raise Exception("Stack is empty.")
            node = self.head
            self.head = node.next
            return node
        except Exception as exc:
            print(exc[0])
    
    def peek(self) -> Node:
        try:
            if self.is_empty():
                raise Exception("Stack is empty.")
            return self.head
        except Exception as exc:
            print(exc[0])